cli/azd/design-notes/environments.md

```diff
@@ -0,0 +1,271 @@
+
+As a UX concern, we may need to handle the "location" parameter in a clever way to reuse whatever location is being set for the template. Ideally we will not have to ask separate questions for "where do you want to store your resources" and "where do you want to store your deployment metadata" even though these are individually configurable.
+
+Note that the target **does not** contain information about a deployment name. Today the name we use for the deployment matches the environment name. We will change this to instead include the *project name* (stored in the name property of `azure.yaml` and the *environment id* to ensure uniqueness. We should also include some per deployment entropy (or perhaps just a timestamp) such that each deployment gets its own deployment object. This should help with some problems we face today where re-using the same deployment object can confuse our status reporting logic, and provides a better view of deployment history.  
+
+Since deployment names are limited to 64 characters, we will need someway to truncate names if we hit this limit.  We should bias towards keeping as much entropy as possible in the strings to ensure uniqueness at a given scope. We should tag these deployment objects with the full environment ID for querying later.
+
+## User Impacting Changes
+
+- We no longer add `AZURE_ENV_NAME`, `AZURE_LOCATION`, `AZURE_SUBSCRIPTION_ID` or `AZURE_PRINCIPAL_ID` to the `.env` file.
+- We no longer support environment variable substitution in the `.parameters.json` file in the `infra` folder, it is used just to configure static values.
+- In our templates we remove the `.parameters.json` files since they are no longer used (all values are user configured during first deployment). (*note: We can choose to keep these files if we think their existence provides value in our templates as a learning tool*)
+- We prompt and validate deployment parameters using the metadata applied in the bicep template.
+
+We can figure out how to rolls these changes out in a smooth enough way (for example, we could take the existence of a environment reference like `${FOO}` in a `.parameters.json` to mean we were on the "old" plan and `azd` can warn and do something sensible (or point folks at an article on how to upgrade).
+
+## Open Questions
+
+> How do we express that a deployment parameter should correspond to the environment id or environment name. Do we introduce types like `azd-env-id` and `azd-env-name` for our metadata on parameters? Should we instead have something like `azd-context` and that passes in a JSON object with well known keys like `id` and `name`?  Is there value in making it easy to flow the environment name (which now will often just be the string "default" into the infrastructure? Why would you use it vs `azd-env-id`?) 
```

Comment:
How do I **easily** associate a ANY parameter with an environment variable?  Right now that is done via parameters.json and env var subs and is easy to see what is happening and explain how it happens.  We shouldn't lose that simplicity.